Compute embodied carbon change during window enhancement.

prsh5175 opened the pull request:
This measure adds IGU to existing windows and computes associated embodied carbon. 

(Branch name may be misleading since we first started with window replacement as the target but changed to enhancement later on. Keeping the branch name original to avoid workflow complexity/confusion.)

**To-do items for this PR:**
- [x] Identify different window layers, identify whether it is a SimpleGlazing object or not, if not, identify each layer and corresponding attributes.
- [x] Add IGU calculations.
- [x] Auto-generate URL strings for each unique layer. 
- [x] Attach the AdditionalProperties object to subsurface (vs building) and store the CO2e value there. We are going to store this number uniquely for each material (as a total value per material). 
- [x] Add one more user argument asking for the API key (@1jhu1 )
- [ ] Clean up and better organize the files. Some need to be gitignored and some need to be deleted. 
- [x] Add industrial EPDs from EC3. Industrial GWP values are reported as per unit mass, whereas we only get volume from an OS model. We will pull the density information from OS model. 
- [x] Correctly attach the GWP values to AdditionalProperties object. We should address this at the end.

prsh5175 committed:
Initial commit. Simply copied a measure as a starting point.

prsh5175 committed:
Merge branch 'SCOPE-measures' into replace-windows-EC

prsh5175 committed:
Synced with latest parent branch.

Commit:
update the ReplaceExteriorWindowConstruction_Test.rb with EC3 API call

Commit:
API Call edits

prsh5175 committed:
Merge branch 'JY_SCOPE-measures' into replace-windows-EC

prsh5175 committed:
Added minor comments and cleanup.

Commit:
edits

Commit:
change EPD location to US and Canada

prsh5175 committed:
Measure name, description etc. updated and packaged EC3_lookup in a separate helper file.

prsh5175 committed:
Test design document added.

Commit:
Updated version

Commit:
edits on unit conversion

Commit:
calculate GWP mean

kflemin committed:
fix dependencies

Commit:
prep 0.11.1

kflemin committed:
Merge pull request #64 from NREL/os39-fix

Update dependencies for OS 3.9

kflemin committed:
remove addressable

kflemin committed:
Merge pull request #66 from NREL/os39-fix

remove addressable

prsh5175 committed:
Test commit.

prsh5175 committed:
Test commit

prsh5175 committed:
Initial attempt to add IGU frame gwp (needs bugfixes).

Commit:
bug to be fixed with configparser, unable to read url from config.ini

prsh5175 committed:
Temporary commit.

prsh5175 committed:
Merge branch 'replace-windows-EC' of https://github.com/NREL/openstudio-ee-gem into replace-windows-EC

prsh5175 committed:
Measure name typo correction and gitignored __pycache__

prsh5175 commented:
measure.rb is experimental to investigate correspondence across Python and Ruby measures. It will be deleted later.

prsh5175 committed:
Added a reporting line and test updates.

Commit:
translate ruby measure to python in ReplaceExteriorWindowConstruction

Commit:
Merge branch 'replace-windows-EC' of https://github.com/NREL/openstudio-ee-gem into replace-windows-EC

DavidGoldwasser committed:
removing measure.rb from python measure

DavidGoldwasser committed:
adding some logic to measure to step through subsurfaces

prsh5175 committed:
Python measure and tests updated.

prsh5175 committed:
gitignore, measure.py updated

prsh5175 committed:
Commit for debug.

prsh5175 committed:
perimeter now computed from model vs user argument. Both measure and test file updated. Additional temporary file created to list out all model attributes for faster search.

prsh5175 committed:
Reduced the user-defined arguments.

prsh5175 committed:
layered construction now working.

Commit:
url address extracted using configparser is causing error, can only extract API_TOKEN from config.ini

Commit:
API token and url address extracted using configparser is causing error, temporarily stopped using configparser

Commit:
Revisions: Line 151-178 in measure.py for WindowEnhancement, needs Prateek to review

prsh5175 commented on lib/measures/window_enhancement/resources/EC3_lookup.py:
```diff
@@ -122,9 +123,9 @@ def parse_gwp_data(epd: Dict[str, Any]) -> Dict[str, Any]:
-    parsed_data["gwp_per_unit_mass"] = f"{gwp_per_unit_mass} kg CO2 e/kg" if gwp_per_unit_mass else "Not specified"
+    parsed_data["gwp_per_unit_volume (kg CO2 eq/m3)"] = gwp_per_unit_volume
+    parsed_data["gwp_per_unit_area (kg CO2 eq/m2)"] = gwp_per_unit_area
+    parsed_data["gwp_per_unit_mass (kg CO2 eq/kg)"] = gwp_per_unit_mass if gwp_per_unit_mass else "Not specified"
```
If gwp_per_unit_mass could be 0 and that is a valid value, this check will incorrectly replace it with "Not specified". To fix this, we could use: 
parsed_data["gwp_per_unit_mass (kg CO2 eq/kg)"] = gwp_per_unit_mass if gwp_per_unit_mass is not None else "Not specified"

1jhu1 replied:
Thanks for checking my commit! I think what we have is fine here. With or without "is not None", it should work the same. What I did is just move the unit text (e.g.,"kg CO2 eq/m3)" into the key of dictionary object parsed_data, so that we don't need to extract numeric values. The part I hope you can help with reviewing are Lines 151-178 in measure.py of WindowEnhancement.

prsh5175 commented on lib/measures/window_enhancement/measure.py:
```diff
@@ -144,13 +148,34 @@ def run(self, model: openstudio.model.Model, runner: openstudio.measure.OSRunner
-            gwp_per_volume = calculate_gwp_per_volume(gwp, declared_unit)
-            runner.registerInfo(f"GWP per volume: {gwp_per_volume:.2f} kgCO2e/m3.")
+            ### revision starts ###
+            wframe_data = fetch_epd_data(URLS["wframe"])
```
This is good progress. Right now we have just a single window frame material identified through the EC3_lookup.py (hard coded). If a model has more number of materials, instead of specifying the exact key of the URLS dictionary, perhaps we could first count the number of different materials encountered in the model. Then we identify the name of each of those materials and then perform an EC3 lookup, that would populate a dictionary (or a JSON file) for the looked up values. Then we can put this line in a loop for the number of different materials encountered.

1jhu1 replied:
Yes, that's also what I expected. Not only URL dictionary, but also need something like a dropdown list for users to choose from multiple EPDs returned. We can discuss on this further. Do you think it's okay that we temporarily use two mean values for GWP of window frame and insulated glass unit so as to get things running successfully?

prsh5175 replied:
OK for now, but we definitely want to separately calculate these and add them up, eventually.

prsh5175 commented on lib/measures/window_enhancement/measure.py:
```diff
@@ -7,6 +7,10 @@
 import typing
 from pathlib import Path
 from resources.EC3_lookup import calculate_gwp_per_volume  # Ensure this function exists and works correctly
```
This line can now be deleted.

prsh5175 commented on lib/measures/window_enhancement/measure.py:
```diff
@@ -144,13 +148,34 @@ def run(self, model: openstudio.model.Model, runner: openstudio.measure.OSRunner
 
-        total_embodied_carbon = total_window_frame_volume * gwp_per_volume
+        total_embodied_carbon = total_window_frame_volume * wframe_mean_gwp_per_volume
+        ### revision ends above ### After Prateek reviews the revisions here, we will then modify other sections accordingly
```
Let's discuss about taking the average value. But other than that, rest of the Method looks OK to me.

prsh5175 replied:
Resolution: averaging approach is fine. We will also report out a range of GWPs.

prsh5175 committed:
Getting rid of some unused lines.

prsh5175 committed:
Minor updates, comments updated, redundant lines deleted. 'total_embodied_carbon' changed to 'frame_embodied_carbon' since IGU calculations are still remaining (added corresponding comments).

prsh5175 committed:
Test commit to check inter-operability with sub-branch.

Commit:
new edits in EC3_lookup.py (new function for generating url) and measure.py for window enhancement (new choice arguments and get min/max/mean gwp values for embodeid carbon calculation)

Commit:
connect generate_url with user arguments; add median gwp option; unable to include industry EPD, need help from EC3

Commit:
saved api token to onedrive folder; code for generating EPD data of igu and window frame is ready

Commit:
consider product lifetime when calculating total embodied carbon; revise lines 371-380 later with the help of Prateek

Commit:
Add notations for subsurface looping; Add edits for window material's volume calcualtion

Commit:
Error free version that can calculate volume of window frame and igu

Commit:
Error free version that can calculate total embodied carbon of window construction

prsh5175 committed:
tests updated and config.ini working for EC3_lookup

prsh5175 committed:
Merge branch 'replace-windows-EC' of https://github.com/NREL/openstudio-ee-gem into replace-windows-EC

prsh5175 committed:
`apply_measure.py` portion added to `test_window_enhancement.py`

prsh5175 committed:
Dummy additional_properties object added.

prsh5175 committed:
'del model' line added at the end of apply_measure.py for memory leak management.

prsh5175 committed:
Saving the modified model to check that additionalProperties object is added to the OSM.

prsh5175 commented on lib/measures/window_enhancement/resources/EC3_lookup.py:
```diff
@@ -0,0 +1,218 @@
+
+config = configparser.ConfigParser()
+config.read(config_path)
+API_TOKEN= config["EC3_API_TOKEN"]["API_TOKEN"]
```
The new plan is to receive the API_TOKEN as a user argument that the user manually needs to input.

prsh5175 commented on lib/measures/window_enhancement/resources/EC3_lookup.py:
```diff
@@ -0,0 +1,218 @@
+material_category = {
+                     "glazing":["InsulatingGlazingUnits"],
+                     "extrusions":["AluminiumExtrusions"]
+                     }
```
material_category dictionary is repeated twice. One can be deleted.

prsh5175 commented on lib/measures/window_enhancement/model_attributes.txt:
This file can be deleted.

prsh5175 commented on lib/measures/window_enhancement/temp.py:
This file can be deleted.

prsh5175 commented on lib/measures/window_enhancement/tests/__pycache__/test_window_enhancement.cpython-39-pytest-7.1.2.pyc:
__pycache__ folder needs to be gitignored.

prsh5175 commented on lib/measures/ReplaceExteriorWindowConstruction/measure.py:
This diff can be deleted. We are not touching any other measure than the one we are concerned with.

prsh5175 commented on lib/measures/ReplaceExteriorWindowConstruction/tests/ReplaceExteriorWindowConstruction_Test.py:
This diff can be deleted. We should only be concerned with window_enhancement measure in this PR.

prsh5175 commented on lib/measures/window_enhancement/README.md:
Needs update.

prsh5175 commented on lib/measures/window_enhancement/measure.py:
```diff
@@ -0,0 +1,431 @@
+        args.append(frame_cross_section_area)
+
+        igu_thickness = openstudio.measure.OSArgument.makeDoubleArgument("igu_thickness", True)
+        igu_thickness.setDisplayName("Thickness of Insulating Glazing Unit (m)")
```
thickness entry in mm instead of m?

Commit:
Add FrameAndDivider to example model for window frame calculations; Restructure looping to assign additional properties: embodied carbon of window frame is assigned to "Window Property Frame And Divider 1", embodeid carbon of igu is assigned to "Exterior Window"

prsh5175 commented:
@1jhu1 please pull from the `develop` branch occasionally to keep it in sync with the latest version.

Commit:
Add arguments: api_key, epd_type; edit functions: parse_product_epd and parse_industrial_epd

Commit:
Merge branch 'develop' of https://github.com/NREL/openstudio-ee-gem into replace-windows-EC

prsh5175 commented on lib/measures/window_enhancement/measure.py:
```diff
@@ -0,0 +1,433 @@
+                runner.registerInfo(f"Error processing argument: {arg_name} - {str(e)}")
+        
+        # Check if numeric values are reasonable
+        if num_panes == 0 or num_panes > 2:
```
Should these checks be integrated to tests instead of the main measure?

1jhu1 replied:
I am not clear about the difference. I added these checks here simply by following other ruby measures in the repo. The meaning here is, if these numeric values are not reasonable, users are required to modify these values. Otherwise, the measure can not calculate the total embodied carbon.

prsh5175 replied:
Ok, let's keep them here then.

prsh5175 commented on lib/measures/window_enhancement/measure.py:
```diff
@@ -0,0 +1,433 @@
+                runner.registerInfo(f"Material: {opaque_material.nameString()}, Thickness: {opaque_material.thickness()} m")
+                igu_thickness += opaque_material.thickness()
+            else:
+                igu_thickness = 0.006 # assign a value if the window construction has no thickness
```
Noting here that we have hard-coded this for now. More complex logic will be needed here later.

1jhu1 replied:
Do you think if we can modify the example model by changing simple glazing to glazing window material: Clear 3mm? Doing so we can then retrieve thickness from openstudio model.

prsh5175 commented on lib/measures/window_enhancement/measure.py:
```diff
@@ -0,0 +1,433 @@
+                frame_width = frame.frameWidth()
+                runner.registerInfo(f"SubSurface: {subsurface.nameString()}'s Frame: {frame_name}, Frame Width: {frame_width} m")
+                window_area = frame_width*frame_width
+                frame_perimeter = frame_width*4
```
The assumption here is that the windows are square. Why did we deviate here from our previous approach of identifying the vertices and calculating the perimeter from the vertices?

1jhu1 replied:
Good question! Because only frame width is provided and this is not a subsurface object. This is a FrameAndDivider object, can we still use the vertices?

prsh5175 commented on lib/measures/window_enhancement/measure.py:
```diff
@@ -0,0 +1,433 @@
+                            building_material_names.append(material.nameString())
+                            runner.registerInfo(f"Layer {i+1}: {material.nameString()}")
+                    else:
+                        runner.registerWarning(f"Construction of Sub-surface {sub_surface_name} is not a LayeredConstruction.")
```
Instead of exiting the loop after this, we need to check if it is a SimpleGlazing object and add more complexity to identify a suitable candidate. TBD later.

1jhu1 replied:
I am not sure if I understand your comment... If you try to run apply_method.py. You can see the printed message showing that there is only one layer for this layered construction, and the only layer is Simple Glazing. We can discuss more if this is not what you are asking.

INFO: Construction of Sub Surface 2 is a LayeredConstruction.
INFO: Layer 1: Simple Glazing

Commit:
Increase example model complexity; Reorganize EC3_lookup; Revise window frame calculations, assign embodied carbon to corresponding subsurface object

Commit:
minor revisions and clean up

prsh5175 committed:
Deleted non essential file.

Commit:
handle triple pane, include gap layer thickness for igu calculation

Commit:
Merge branch 'replace-windows-EC' of https://github.com/NREL/openstudio-ee-gem into replace-windows-EC

prsh5175 committed:
Minor debug.

Commit:
deleted 'single_value' option as it is not necessary; revised keyword for precast concrete EPD search

Commit:
Merge branch 'replace-windows-EC' of https://github.com/NREL/openstudio-ee-gem into replace-windows-EC

prsh5175 committed:
Two example models created with slightly different window properties, measures applied to observe delta EC, ran both models for delta OC

prsh5175 committed:
Re-run of both models.

DavidGoldwasser committed:
initial commit for 3.10 prep

DavidGoldwasser committed:
update gemspec to match common measures

DavidGoldwasser committed:
commenting to problematic test related to sizing issue

anchapin committed:
Add GitHub workflow for OpenStudio testing

Set up a CI workflow that runs tests in an OpenStudio 3.10.0 container. The workflow checks out the repository, installs dependencies, runs spec tests and measure tests, then uploads test results to an S3 bucket for dashboard visualization.

DavidGoldwasser committed:
Merge pull request #70 from NREL/3.10.0_prep

3.10.0 prep

DavidGoldwasser committed:
removing inadvertent measure

DavidGoldwasser committed:
progress update. Tests not passing, fix or merge_measure to tests

DavidGoldwasser committed:
2 tests running, update other tests and then merge

DavidGoldwasser committed:
Enabled unit tests removed apply_measure

need to add additional tests when epd_type is Indusstry

DavidGoldwasser committed:
Merge branch 'SCOPE-measures' into replace-windows-EC
